hail/src/main/scala/is/hail/expr/ir/functions/LocusFunctions.scala

```diff
@@ -52,43 +48,45 @@ object LocusFunctions extends RegistryFunctions {
+        val tuple = Code.invokeScalaObject[Locus, IndexedSeq[String], (Locus, IndexedSeq[String])](VariantMethods.getClass, "minRep", locus, alleles)
 
-      val newLocus = Code.checkcast[Locus](returnTuple.load().get[java.lang.Object]("_1"))
-      val newAlleles = Code.checkcast[IndexedSeq[String]](returnTuple.load().get[java.lang.Object]("_2"))
+        val newLocus = Code.checkcast[Locus](returnTuple.load().get[java.lang.Object]("_1"))
+        val newAlleles = Code.checkcast[IndexedSeq[String]](returnTuple.load().get[java.lang.Object]("_2"))
 
-      val srvb = new StagedRegionValueBuilder(r, PTuple(FastIndexedSeq(tv("T").t.physicalType, PArray(PString()))))
-      Code(
-        returnTuple := tuple,
-        srvb.start(),
-        srvb.addBaseStruct(types.coerce[PBaseStruct](tv("T").t.fundamentalType.physicalType), { locusBuilder =>
-          Code(
-            locusBuilder.start(),
-            locusBuilder.addString(newLocus.invoke[String]("contig")),
-            locusBuilder.advance(),
-            locusBuilder.addInt(newLocus.invoke[Int]("position")))
-        }),
-        srvb.advance(),
-        srvb.addArray(PArray(PString()), { allelesBuilder =>
-          Code(
-            allelesBuilder.start(newAlleles.invoke[Int]("size")),
-            Code.whileLoop(allelesBuilder.arrayIdx < newAlleles.invoke[Int]("size"),
-              allelesBuilder.addString(Code.checkcast[String](newAlleles.invoke[Int, java.lang.Object]("apply", allelesBuilder.arrayIdx))),
-              allelesBuilder.advance()))
-        }),
-        srvb.offset)
+        val newLocusT = PLocus(locusT.rg)
+        val newAllelesT = PArray(PString())
+        val srvb = new StagedRegionValueBuilder(r, PTuple(FastIndexedSeq(newLocusT, newAllelesT)))
+        Code(
+          returnTuple := tuple,
+          srvb.start(),
+          srvb.addBaseStruct(newLocusT.fundamentalType.asInstanceOf[PStruct], { locusBuilder =>
+            Code(
+              locusBuilder.start(),
+              locusBuilder.addString(newLocus.invoke[String]("contig")),
+              locusBuilder.advance(),
+              locusBuilder.addInt(newLocus.invoke[Int]("position")))
+          }),
+          srvb.advance(),
+          srvb.addArray(newAllelesT, { allelesBuilder =>
+            Code(
+              allelesBuilder.start(newAlleles.invoke[Int]("size")),
+              Code.whileLoop(allelesBuilder.arrayIdx < newAlleles.invoke[Int]("size"),
+                allelesBuilder.addString(Code.checkcast[String](newAlleles.invoke[Int, java.lang.Object]("apply", allelesBuilder.arrayIdx))),
+                allelesBuilder.advance()))
+          }),
+          srvb.offset)
     }
 
     registerCode("locus_windows_per_contig", TArray(TArray(TFloat64())), TFloat64(), TTuple(TArray(TInt32()), TArray(TInt32()))) {
-      (r: EmitRegion, coords: Code[Long], radius: Code[Double]) =>
+      case (r: EmitRegion, (groupedT: PArray, coords: Code[Long]), (radiusT: PFloat64, radius: Code[Double])) =>
```

Comment:
Should this be `groupedAt: PArray[PArray[PFloat64]]`?

Reply:
That's not valid syntax -- the Hail types won't work as scala types that way unfortunately. But semantically that's what it is.

Reply:
Types aren't parameterized in Scala's type system -- it's `PArray(PInt32())`, which is a class, not `PArray[PInt32]`. So we can't use pattern matching in this way.